core/src/main/java/com/datastax/oss/driver/internal/core/session/DefaultSession.java

```diff
@@ -94,14 +94,30 @@
 
   private DefaultSession(InternalDriverContext context, Set<InetSocketAddress> contactPoints) {
     LOG.debug("Creating new session {}", context.getSessionName());
-    this.adminExecutor = context.getNettyOptions().adminEventExecutorGroup().next();
-    this.context = context;
-    this.singleThreaded = new SingleThreaded(context, contactPoints);
-    this.metadataManager = context.getMetadataManager();
-    this.processorRegistry = context.getRequestProcessorRegistry();
-    this.poolManager = context.getPoolManager();
     this.logPrefix = context.getSessionName();
-    this.metricUpdater = context.getMetricsFactory().getSessionUpdater();
+    this.adminExecutor = context.getNettyOptions().adminEventExecutorGroup().next();
```

Comment:
this method call cannot throw an exception? I see that we enclosed in a `try` everything besides that line

Reply:
I suppose it could throw an exception, but it would only be in the DefaultNettyOptions constructor. If the constructor doesn't throw an exception, `adminEventExecutor()` simply returns the EventLoopGroup instance created in the constructor, and `next()` just returns the next executor in the group.

If an exception is thrown in the constructor, it will fall out of this `DefaultSession` constructor and bubble up the stack, just as would any Exception that would be thrown from any of the methods moved into the try block.
The reason this line is not in the try block is that if it throws an Exception, there won't be a NettyOptions instance created on the Context (it's lazily created on the first call to `getNettyOptions()` on the context). If there is no NettyOptions instance, then there isn't a need to call `onClose()` to free resources.

The idea here is that if any of these methods (including this line) throws an Exception, it should propagate up. However, if a NettyOptions is successfully created, but something else fails, we need to free the resources created in the NettyOptions before the Exception propagates.

Reply:
Thanks for a detailed exmplanation

Reply:
👍 
In theory the `DefaultNettyOptions` constructor could throw after it created the timer, but we wouldn't have a way to stop the timer from here anyway.
And given that it's the last thing done in that constructor, there's little chance that this could happen.